python-checks/src/main/java/org/sonar/python/checks/InstanceMethodSelfAsFirstCheck.java

```diff
@@ -0,0 +1,197 @@
+          }
+        }
+
+        return false;
+      });
+    }
+  }
+
+  private static boolean isUsedAsOldStyleClassMethod(ClassDef parent, Symbol symbol) {
+    OldStyleClassMethodVisitor visitor = new OldStyleClassMethodVisitor(symbol);
+    parent.accept(visitor);
+
+    return visitor.isOldStyleClassMethod;
+  }
+
+  private boolean isRelevantMethod(ClassDef classDef, FunctionDef functionDef) {
+    // Skip some known special methods
+    if (EXCEPTIONS.contains(functionDef.name().name())) {
+      return false;
+    }
+
+    // Skip if the class has a @classmethod or @staticmethod decorator
+    if (functionDef.decorators().stream().anyMatch(this::isNonInstanceMethodDecorator)) {
+      return false;
+    }
+
+    Symbol functionSymbol = functionDef.name().symbol();
+    if (functionSymbol == null) {
+      return false;
+    }
+
+    ClassSymbol classSymbol = TreeUtils.getClassSymbolFromDef(classDef);
+    if (classSymbol != null && classSymbol.isOrExtends("zope.interface.Interface")) {
+      return false;
+    }
+
+    return !isUsedAsOldStyleClassMethod(classDef, functionSymbol);
+  }
+
+  private void handleFunctionDef(SubscriptionContext ctx, ClassDef classDef, FunctionDef functionDef) {
+    ParameterList parameterList = functionDef.parameters();
+    if (parameterList == null || parameterList.all().isEmpty()) {
+      return;
+    }
+
+    AnyParameter first = parameterList.all().get(0);
+    if (first instanceof Parameter) {
+      Parameter firstParameter = (Parameter) first;
+      Name paramName = firstParameter.name();
+      if (paramName == null) {
+        return;
+      }
+
+      if (firstParameter.starToken() != null) {
+        return;
+      }
+
+      if (!ALLOWED_NAMES.contains(paramName.name()) && isRelevantMethod(classDef, functionDef)) {
+        ctx.addIssue(first, String.format("Rename \"%s\" to \"self\" or add the missing \"self\" parameter.", paramName.name()));
+      }
+    }
+  }
+
+  private static boolean isRelevantClass(ClassDef classDef) {
+    ClassSymbol classSymbol = TreeUtils.getClassSymbolFromDef(classDef);
+    if (classSymbol == null) {
+      return false;
+    }
+
+    // Do not raise on nested classes
+    Tree possibleParent = TreeUtils.firstAncestorOfKind(classDef, Tree.Kind.CLASSDEF);
+    if (possibleParent != null) {
+      return false;
+    }
+    
+    // Do not raise on meta-classes
+    return !classSymbol.isOrExtends("type");
+  }
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CLASSDEF, ctx -> {
+      ClassDef classDef = (ClassDef) ctx.syntaxNode();
+      if (!isRelevantClass(classDef)) {
+        return;
+      }
+
+      List<FunctionDef> functions = classDef.body().statements().stream()
+        .filter(s -> s.is(Tree.Kind.FUNCDEF))
+        .map(s -> (FunctionDef) s)
+        .collect(Collectors.toList());
+
+      for (FunctionDef functionDef : functions) {
```

Comment:
I think you can avoid collecting to List at line 190 and do a "foreach" instead.